pkg/controller/node/ipam/cloud_cidr_allocator.go

```diff
@@ -42,84 +45,149 @@ import (
+	}
 
 	gceCloud, ok := cloud.(*gce.GCECloud)
 	if !ok {
-		err = fmt.Errorf("cloudCIDRAllocator does not support %v provider", cloud.ProviderName())
-		return
+		err := fmt.Errorf("cloudCIDRAllocator does not support %v provider", cloud.ProviderName())
+		return nil, err
 	}
 
-	ca = &cloudCIDRAllocator{
-		client: client,
-		cloud:  gceCloud,
-		recorder: record.NewBroadcaster().NewRecorder(
-			api.Scheme,
-			v1.EventSource{Component: "cidrAllocator"}),
+	ca := &cloudCIDRAllocator{
+		client:                client,
+		cloud:                 gceCloud,
+		nodeCIDRUpdateChannel: make(chan nodeAndCIDR, cidrUpdateQueueSize),
+		recorder:              recorder,
+		nodesInProcessing:     sets.NewString(),
+	}
+
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		go func(stopChan <-chan struct{}) {
+			for {
+				select {
+				case workItem, ok := <-ca.nodeCIDRUpdateChannel:
+					if !ok {
+						glog.Warning("NodeCIDRUpdateChannel read returned false.")
+						return
+					}
+					ca.updateCIDRAllocation(workItem)
+				case <-stopChan:
+					return
+				}
+			}
+		}(wait.NeverStop)
 	}
 
 	glog.V(0).Infof("Using cloud CIDR allocator (provider: %v)", cloud.ProviderName())
+	return ca, nil
+}
 
-	return
+func (ca *cloudCIDRAllocator) insertNodeToProcessing(nodeName string) bool {
+	ca.Lock()
+	defer ca.Unlock()
+	if ca.nodesInProcessing.Has(nodeName) {
+		return false
+	}
+	ca.nodesInProcessing.Insert(nodeName)
+	return true
 }
 
-func (ca *cloudCIDRAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
-	glog.V(2).Infof("Updating PodCIDR for node %v", node.Name)
+func (ca *cloudCIDRAllocator) removeNodeFromProcessing(nodeName string) {
+	ca.Lock()
+	defer ca.Unlock()
+	ca.nodesInProcessing.Delete(nodeName)
+}
 
+// WARNING: If you're adding any return calls or defer any more work from this
+// function you have to handle correctly nodesInProcessing.
+func (ca *cloudCIDRAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	if node == nil {
+		return nil
+	}
+	if !ca.insertNodeToProcessing(node.Name) {
+		glog.V(2).Infof("Node %v is already in a process of CIDR assignment.", node.Name)
+		return nil
+	}
 	cidrs, err := ca.cloud.AliasRanges(types.NodeName(node.Name))
-
 	if err != nil {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-
 	if len(cidrs) == 0 {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
-		glog.V(2).Infof("Node %v has no CIDRs", node.Name)
-		return fmt.Errorf("failed to allocate cidr (none exist)")
+		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
 
-	node, err = ca.client.Core().Nodes().Get(node.Name, metav1.GetOptions{})
-	if err != nil {
-		glog.Errorf("Could not get Node object from Kubernetes: %v", err)
-		return err
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, cidrs[0])
```

Comment:
V(4) is fine, no need to go all the way to 10

Reply:
Done.. Just an artifact from the corresponding code in range allocator.